compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp

```diff
@@ -57,127 +86,273 @@ static bool areFusableOps(MLIRContext *context, Operation *producerOp,
+    return success();
+  }
+};
+
+static FailureOr<unsigned> fuseMultiUseProducers(Operation *funcOp,
+                                                 MLIRContext *context,
+                                                 DominanceInfo &dominanceInfo) {
+  // Try fusion of operations when producer has multiple uses.
+  // 1. Walk the function in pre-order.
+  // 2. Check if a `linalg.generic` op has a consumer `linalg.generic` op
+  //    that dominates all uses of the producer op. Then fuse the producer
+  //    consumer
+  unsigned numCandidates = 0;
+  OpBuilder builder(context);
+  funcOp->walk<WalkOrder::PreOrder>([&](linalg::GenericOp genericOp) {
+    auto consumerAttrName =
+        FuseElementwiseOpsWithMultipleUses::getConsumerAttributeName();
+    auto producerAttrName =
+        FuseElementwiseOpsWithMultipleUses::getProducerAttributeName();
+    if (genericOp->hasAttr(consumerAttrName) ||
+        genericOp->hasAttr(producerAttrName)) {
+      return;
+    }
+
+    Optional<OpOperand *> fusableUse = getFusableUse(genericOp, dominanceInfo);
+    if (!fusableUse) return;
+    if (!linalg::areElementwiseOpsFusable(fusableUse.getValue())) return;
+
+    Operation *consumer = fusableUse.getValue()->getOwner();
+    genericOp->setAttr(producerAttrName,
+                       builder.getI64IntegerAttr(numCandidates));
+    consumer->setAttr(consumerAttrName,
+                      builder.getI64IntegerAttr(numCandidates));
+    numCandidates++;
+    return;
+  });
+  LLVM_DEBUG({
+    llvm::dbgs() << "Num of multiuse fusable candidates : " << numCandidates
+                 << "\n";
+    funcOp->print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+  });
+  RewritePatternSet fusionPatterns(context);
+  fusionPatterns.insert<FuseElementwiseOpsWithMultipleUses>(context);
+  linalg::GenericOp::getCanonicalizationPatterns(fusionPatterns, context);
+  if (failed(applyPatternsAndFoldGreedily(funcOp->getRegions(),
+                                          std::move(fusionPatterns)))) {
+    return funcOp->emitOpError("multi use producer -> consumer fusion failed");
+  }
+  return numCandidates;
+}
+
 /// Pass to fuse linalg on tensor operations as well as fusion of hal.interface*
 /// operations with linalg.tensor_reshape operation.
-struct FusionOfTensorOpsPass
+class FusionOfTensorOpsPass
     : public FusionOfTensorOpsBase<FusionOfTensorOpsPass> {
+ public:
   void getDependentDialects(DialectRegistry &registry) const override {
     registry.insert<AffineDialect, linalg::LinalgDialect, math::MathDialect>();
   }
+  FusionOfTensorOpsPass(bool fuseMultiUse, unsigned multiUseFusionIteration)
+      : fuseMultiUse(fuseMultiUse),
+        multiUseFusionIteration(multiUseFusionIteration) {}
+  FusionOfTensorOpsPass(const FusionOfTensorOpsPass &pass)
+      : fuseMultiUse(pass.fuseMultiUse),
+        multiUseFusionIteration(pass.multiUseFusionIteration) {}
+
+  LogicalResult initializeOptions(StringRef options) override {
```

Comment:
Huh, interesting. Are you sure this is not the other way round. IIUC the options added in tblgen are already members of the base class. You can use the values passed in the constructor to assign the values to the options, and then you can just leave the base class `initializeOptions` as is and not override them. Within the code you can use the option directly as a `bool` or `unsigned` value....

Reply:
SG. Thx.